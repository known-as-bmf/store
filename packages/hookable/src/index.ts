export { createHookable } from './hookable';
export { every, once } from './hook-subscribers';
export {
  Hook,
  HookSubscriber,
  Hooks,
  TapInput,
  TapOutput,
  TransformInput,
  TransformOutput,
} from './types';
