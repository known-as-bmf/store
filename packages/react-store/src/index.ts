export { useStore } from './useStore';
export { useNewStore } from './useNewStore';
