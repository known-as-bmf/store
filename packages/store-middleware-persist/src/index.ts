export { PersistOptions } from './types';
export { persistMiddleware } from './persist';
