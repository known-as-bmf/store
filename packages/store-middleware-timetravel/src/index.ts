export { TimetravelOptions } from './types';
export { timetravelMiddleware, undo, redo, derefHistory } from './timetravel';
