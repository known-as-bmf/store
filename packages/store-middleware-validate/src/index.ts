export { Validator } from './types';
export { validatorMiddleware } from './validator';
