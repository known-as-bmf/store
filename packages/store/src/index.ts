export {
  Hooks,
  Middleware,
  Selector,
  StateChangedEvent,
  Store,
  SubscriptionCallback,
} from './types';
export { deref, of, set, subscribe, swap } from './store';
export { composeMiddlewares, pipeMiddlewares } from './composeMiddlewares';
