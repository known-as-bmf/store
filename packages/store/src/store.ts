import { produce } from 'immer';
import { createHookable } from '@known-as-bmf/hookable';

import { errors } from './errors';
import {
  SubscriptionCallback,
  Selector,
  StateChangedEvent,
  Hooks,
  Store,
  Middleware,
} from './types';

// Internal symbols used by the store instance.
/**
 * Current subscriptions to store changes.
 */
const SUBSCRIPTIONS = Symbol();
/**
 * Current state of the store.
 */
const STATE = Symbol();
/**
 * State transformation (middleware).
 */
const SET_STATE = Symbol();

/**
 * The store class.
 */
class StoreImpl<S> implements Store<S> {
  public declare [SUBSCRIPTIONS]: Set<Subscription<S, unknown>>;
  public declare [STATE]: S;
  public declare [SET_STATE]: (state: S) => S;
}

/**
 * Internal subscription type.
 */
type Subscription<S, R> = [Selector<S, R>, SubscriptionCallback<S>];

/**
 * Check if a value is a Store instance.
 * @param input The value to check.
 */
function isStore<S = unknown>(input: unknown): input is StoreImpl<S> {
  return input instanceof StoreImpl;
}

function assertStore<S>(store: Store<S>): asserts store is StoreImpl<S> {
  if (!isStore(store)) {
    throw new TypeError(errors.invalidStore);
  }
}

function getState<S>(store: StoreImpl<S>): S {
  const { [STATE]: state } = store;
  return state;
}

function setState<S>(store: StoreImpl<S>, state: S): void | never {
  store[SET_STATE](state);
}

/**
 * Notify all subscribers to the store that the state has changed.
 * If a selector was provises at subscription,
 * only notify if the desired value has changed.
 * @param store The store instance.
 * @param event The change event.
 */
function notifySubscribers<S>(
  store: StoreImpl<S>,
  event: StateChangedEvent<S>
): void {
  store[SUBSCRIPTIONS].forEach(([selector, callback]) => {
    if (selector(event.previous) !== selector(event.current)) {
      callback(event);
    }
  });
}

/**
 * Creates a store.
 * @param initialState The initial value of the state.
 * @param middleware Middleware to use for this store. You can compose multiple
 * middlewares with `composeMiddlewares` and `pipeMiddlewares`.
 */
export function of<S>(initialState: S, middleware?: Middleware<S>): Store<S> {
  const store: StoreImpl<S> = Object.defineProperties(
    Object.create(StoreImpl.prototype),
    {
      [SUBSCRIPTIONS]: {
        value: new Set<Subscription<S, unknown>>(),
        writable: true,
      },
      [STATE]: { value: undefined, writable: true },
    }
  ) as StoreImpl<S>;

  const setStateHook = createHookable(
    (s: S): S => {
      store[STATE] = s;
      return s;
    }
  );

  Object.defineProperty(store, SET_STATE, { value: setStateHook });

  const hooks: Hooks<S> = {
    stateDidChange: setStateHook.leave,
    stateWillChange: setStateHook.enter,
    transformState: setStateHook.transformInput,
  };

  if (middleware) {
    middleware(store, hooks);
  }

  setState(store, initialState);

  return store;
}

/**
 * Returns the current state of a store.
 * @param store The store you want to get the current state from.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function deref<S>(store: Store<S>): S {
  assertStore(store);

  return getState(store);
}

/**
 * Changes the state of a store using a function.
 * @param store The store of which you want to change the state.
 * @param mutationFn The function used to compute the value of the future state.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 * @throws {Error} if the new state does not pass validation.
 */
export function swap<S>(store: Store<S>, mutationFn: (current: S) => S): void {
  assertStore(store);

  const previous = getState(store);
  const current = produce(previous, mutationFn) as S;

  setState(store, current);
  notifySubscribers(store, { previous, current });
}

/**
 * Changes the state of a store with a new one.
 * @param store The store of which you want to change the state.
 * @param current The new state.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 * @throws {Error} if the new state does not pass validation.
 */
export function set<S>(store: Store<S>, current: S): void {
  assertStore(store);

  const previous = getState(store);

  setState(store, current);
  notifySubscribers(store, { previous, current });
}

/**
 * Subscribes to state changes.
 * @param store The store you want to subscribe to.
 * @param callback The function to call when the state changes.
 * @returns An unsubscribe function for this specific subscription.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function subscribe<S>(
  store: Store<S>,
  callback: SubscriptionCallback<S>
): () => void;
/**
 * Subscribes to state changes.
 * @param store The store you want to subscribe to.
 * @param callback The function to call when the state changes.
 * @param selector The selector function, narrowing down the part of the state you want to subscribe to.
 * @returns An unsubscribe function for this specific subscription.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function subscribe<S, R>(
  store: Store<S>,
  callback: SubscriptionCallback<S>,
  selector: Selector<S, R>
): () => void;
export function subscribe<S>(
  store: Store<S>,
  callback: SubscriptionCallback<S>,
  selector: Selector<S, unknown> = (s): S => s
): () => void {
  assertStore(store);

  const subscription: Subscription<S, unknown> = [selector, callback];

  store[SUBSCRIPTIONS].add(subscription);
  return () => {
    store[SUBSCRIPTIONS].delete(subscription);
  };
}
