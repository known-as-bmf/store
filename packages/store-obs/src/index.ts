import { Observable, fromEventPattern } from 'rxjs';
import { Store, StateChangedEvent, subscribe } from '@known-as-bmf/store';

/**
 * Creates an observable from the store, emitting every time the store is updated.
 * @param store - The store to observe.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 * @public
 */
export function obs<S>(store: Store<S>): Observable<StateChangedEvent<S>> {
  return fromEventPattern<StateChangedEvent<S>>(
    (callback) => subscribe(store, callback),
    // unsub is the function returned by `subscribe` above
    (_, unsub: () => void) => unsub()
  );
}
