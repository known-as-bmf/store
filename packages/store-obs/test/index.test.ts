import { Observable } from 'rxjs';
import { of, set } from '@known-as-bmf/store';

import { obs } from '../src';

describe('obs', () => {
  it('should be a function', () => {
    expect(obs).toBeDefined();
    expect(obs).toBeInstanceOf(Function);
  });

  it('should return an Observable', () => {
    const store = of('a');

    const obs$ = obs(store);

    expect(obs$).toBeDefined();
    expect(obs$).toBeInstanceOf(Observable);
  });

  it('should emit when the value changes when subscribed', () => {
    const callback = jest.fn();
    const store = of('a');

    const obs$ = obs(store);
    const subscription = obs$.subscribe(callback);

    set(store, 'b');

    expect(callback).toHaveBeenCalledTimes(1);

    subscription.unsubscribe();

    set(store, 'c');

    expect(callback).toHaveBeenCalledTimes(1);
  });

  it('should do nothing if the Observable is not subscribed', () => {
    const callback = jest.fn();
    const store = of('a');

    const obs$ = obs(store);

    set(store, 'b');

    expect(callback).not.toHaveBeenCalled();

    const subscription = obs$.subscribe(callback);

    set(store, 'c');

    expect(callback).toHaveBeenCalledTimes(1);

    subscription.unsubscribe();
  });

  it('should error when subscribing if not given a store', async () => {
    const nextFn = jest.fn();

    obs({}).subscribe(nextFn, (e) => {
      expect(e).toMatchSnapshot();
    });

    obs({ ...of(1) }).subscribe(nextFn, (e) => {
      expect(e).toMatchSnapshot();
    });

    // wait for async operations to finish
    await new Promise((res) => setTimeout(res));

    expect(nextFn).not.toHaveBeenCalled();
  });
});
