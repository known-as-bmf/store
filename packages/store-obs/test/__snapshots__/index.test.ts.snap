// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`obs should error when subscribing if not given a store 1`] = `[TypeError: Provided value is not a valid store instance.]`;

exports[`obs should error when subscribing if not given a store 2`] = `[TypeError: Provided value is not a valid store instance.]`;
